README
= SALOR Hospitality -- The innovative Point of Sale Solution serving the Hospitality Industry

== History

This software product has been in continuous development and used in real-world settings since beginning of 2009. Red (E) Tools Ltd. (http://red-e.eu) is pleased to make it finally available publicly under the MIT License. This product was formerly known as BillGastro (http://billgastro.com), but after the addition of Hotel features in 2012, it was renamed to "SALOR Hospitality" to reflect it's closeness to the hospitality industry as a whole.

== Overview

"SALOR Hospitality" revolutionizes the day to day operation of the Hospitality Industry, bringing businesses large and small into the 21st century. We believe the time has come to integrate smart technology with customer service, taking both to a new level. "SALOR Hospitality" provides insight into every aspect of your daily operations and business across multiple metrics, allowing you to plan performance improvements.

Cutting edge design:
  Appealing design with clean lines and colour schemes for table and staff differentiation, product categories, etc. SalorHospitality looks great on any system!

Flexibility:
  Productive work relationships are the cornerstone of any successful business. SalorHospitality allows you to better coordinate work schedules and responsibilities between staff.

Efficiency:
  SalorHospitality helps you deliver prompt and friendly service to your customers. It is clear, easy and safe to use, optimising workflow from the back office to the kitchen to the restaurant floor.

Extensibility:
  As your business grows, SalorHospitality will grow with you. Its unique system design facilitates seamless enhancements to existing functionality.

Versatility:
  Adapt SalorHospitality to your business by reorganising the table polan, introducing new menus, adding new products and services.

Accessibility:
  Remote access gives you the freedom to run your business anytime from anywhere.

Virtual Restaurant:
  SalorHospitality can extend your restaurant or hotel to your customers' homes. They just take orders online or reserve tables, notifying staff who can prepare in advance.

Multi-Functionality:
  The software multifunctional, allowing you to monitor, export and print customized accounts, statistics and tax reports, update your wholesalers' price lists, easily split receipts, generate professional invoices, and much more.
  
International:
  SALOR Hospitality was translated into 8 languages so far (2012): English, German, French, Spanish, Greek, Russian, Polish, Chinese. Additional languages are currently in the process of translation.

Compatibility:
  "SALOR Hospitality" will run in the browser of any computer and mobile device like Smartphones, so you needn't buy any special devices. However, Red (E) Ltd. can suggest and source hardware optimized for SalorHospitality -- touchscreens, thermal printers, handhelds or tablets.

Many more interestig features:

* Dynamic search function for product database
* Tickets for kitchen and bar
* Instant messaging between all staff
* Immediate updates between all terminals in use
* Enter special customer requests directly at the table -- also in handwriting!
* Compatible with tax schemes in several countries
* Multiple thermal printers
* and much more


== Philosophy of Technology

The applications consists of a lean Ruby-on-Rails backend (Ruby 1.9.3 and Rails 3.2.6) with a MySQL database which is responsible for delivering and storing of data, whereas the the actual Point of Sale interface is entirely controlled by offline Javascript and an offline database. This concept unites the advantages of networking between many handhelds in one store (or many stores of a restaurant/hotel chain) with an amazing responsiveness of the user interface (UI). Because all Javascript and data are residing in the RAM of the browser, a steady connection to the server is not even required. The Javascript application communicates its data via small JSON objects, which only a few dozen bytes per transaction.

Because SALOR Hospitality is based on standard web technologies, it can immediately be used with a wide variety of standard hardware or Operating Systems without hassle: Windows, Linux, OS X, PC's, MAC's, desktops, laptops, handhelds, tablets and smartphones. Just fire up your browser and enter the the URL of your SALOR server. SALOR can distinguish between large and small screens and delivers styles accordingly.

== Maturity

Many hospitality enterprises -- usually high-stress and high-pressure environments -- already enjoy the use of SALOR Hospitality. They have contributed feedback to make SALOR the excellent product it is today.


== Try it live!

You can try SALOR Hospitality immediately. Install Google Chrome, Chromium, Webkit (or any browser based on Webkit), browse to [sh.red-e.eu](http://sh.red-e.eu) and log in with one of the following passwords:

Quick usage hint for the online demo: After you've logged in, you can click on the client logo on the top left of the screen. This opens up the admin menu and enables drag and drop of the tables. When the admin menu is closed, table dragging is disabled.

Warning: The database of the demo application refreshes itself frequently. Do not expect your added content to persist.


== Contact

Dipl.-Ing. (FH) Michael Franzl
Red (E) Tools Ltd.
office@red-e.eu


debian/changelog
salor-hospitality (5.0.25) jessie; urgency=medium

  * deployment optimization

 -- Michael Franzl <office@michaelfranzl.com>  Thu, 17 Sep 2015 09:16:21 +0200

salor-hospitality (5.0.24) jessie; urgency=medium

  * deployment optimization
  * deployment optimization

 -- Michael Franzl <office@michaelfranzl.com>  Thu, 17 Sep 2015 09:01:18 +0200

salor-hospitality (5.0.23) jessie; urgency=medium

  * deployment optimization

 -- Michael Franzl <office@michaelfranzl.com>  Thu, 17 Sep 2015 08:41:41 +0200

salor-hospitality (5.0.22) jessie; urgency=medium

  * better display of longer category names on ordering screen
  * added internationalization support for nigeria
  * Optionally print one ticket per piece
  * fix for saving tax classes
  * Added configuration option to print reduced quantities of items onto tickets
  * Print seconds on tickets for more accuracy
  * load Red E Tools Ltd. logo from the internet
  * added user color to notifications list
  * group notifications list by table
  * fix for passenger configuration

 -- Michael Franzl <office@michaelfranzl.com>  Thu, 17 Sep 2015 08:12:29 +0200

salor-hospitality (5.0.21) jessie; urgency=medium

  * replaced modcoder_excolor colorpicker with colpick
  * show sku input only on workstations, not on mobile devices
  * added option to printers to cut every item in a ticket
  * onscreen keyboard and checkbox javascript fixes for newer jQuery version
  * fixed database seeding problem for options
  * permission changes and migration test
  * fix for deleted payment_methods and cost_centers, report exceptions during settlement finish to technician
  * moved SKU field into better place

 -- Michael Franzl <office@michaelfranzl.com>  Wed, 03 Jun 2015 11:57:47 +0200

salor-hospitality (5.0.20) jessie; urgency=medium

  * removed observer class, app config changes
  * fixed depracated rmagick spelling
  * moved reports into Vendor class, made history creation configurable in config.yml
  * gitignore ignore tmp
  * made Receipt saving to database configurable via config.yml

 -- Michael Franzl <office@michaelfranzl.com>  Tue, 26 May 2015 18:06:07 +0200

salor-hospitality (5.0.19) jessie; urgency=medium

  * postinst fix

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 12:52:00 +0200

salor-hospitality (5.0.18) jessie; urgency=medium

  * updated readme and postinst fix

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 12:40:09 +0200

salor-hospitality (5.0.17) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 12:24:06 +0200

salor-hospitality (5.0.16) jessie; urgency=medium

  * deployment test

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 10:33:53 +0200

salor-hospitality (5.0.15) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:54:58 +0200

salor-hospitality (5.0.14) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:54:03 +0200

salor-hospitality (5.0.13) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:47:49 +0200

salor-hospitality (5.0.12) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:41:55 +0200

salor-hospitality (5.0.11) jessie; urgency=medium

  * test to cache gems in debian package

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:40:28 +0200

salor-hospitality (5.0.10) jessie; urgency=medium

  * fixed and faster order splitting
  * fixed assigning user id
  * javascript fixes
  * fixed payment methods for orders
  * removed debugging logs
  * fix for hand scribe
  * fixes for slideshow

 -- Michael Franzl <office@michaelfranzl.com>  Sun, 24 May 2015 09:20:35 +0200

salor-hospitality (5.0.9) jessie; urgency=medium

  * fix for options

 -- Michael Franzl <office@michaelfranzl.com>  Sat, 23 May 2015 21:10:45 +0200

salor-hospitality (5.0.8) jessie; urgency=medium

  * run seed only once

 -- Michael Franzl <office@michaelfranzl.com>  Sat, 23 May 2015 20:43:15 +0200

salor-hospitality (5.0.7) jessie; urgency=medium

  * removed byebug from deployment Gemfile

 -- Michael Franzl <office@michaelfranzl.com>  Sat, 23 May 2015 20:39:16 +0200

salor-hospitality (5.0.6) jessie; urgency=medium

  * gemfile update
  * strong paramters for bookings working
  * fixed js error for new jquery
  * strong paramters for roles
  * strong paramters for payment methods
  * strong paramters for customers
  * strong paramters for tax classes
  * strong paramters for options
  * strong paramters for categories
  * strong paramters for tables
  * strong paramters for statistic categories
  * JS fix for statistics
  * strong paramters for cost centers
  * strong paramters for cameras
  * more strong paramters
  * fix for surcharge display
  * strong paramters for seasons

 -- Michael Franzl <office@michaelfranzl.com>  Sat, 23 May 2015 20:32:59 +0200

salor-hospitality (5.0.5) jessie; urgency=medium

  * sprockets gem version fix


db/seeds.rb
# Copyright (c) 2012 Red (E) Tools Ltd.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

category_labels = ['Starters','Main Dish','Desserts','Rose Wine','Red Wine','Digestiv'] #,'Alcohol','Coffee','Tea','Tobacco','Beer','Aperitiv','White Wine','Side Dish','Divers']
category_icons = ['starter','maindish','dessert','rosewineglass','redwineglass','digestif'] #,'nonalcoholics','coffee','teapot','cigarette','beer','aperitif','whitewineglass','sidedish','blank']

company_count = 0

if ENV['SEED_MODE'] == 'full'
  puts "SEED_MODE is 'full'"
  countries = ['en','at','fr','es','pl','hu','ru','it','tr','cn','el']
  languages = ['en','gn','fr','es','pl','hu','ru','it','tr','cn','el']
  company_count = 2
elsif ENV['SEED_MODE'] == 'minimal'
  puts "SEED_MODE is 'minimal'"
  countries = ['en']
  languages = ['en']
  company_count = 1
else
  puts "\n\nPlease specify the seed mode with `SEED_MODE={full|minimal} rake db:seed`\n\n"
  Process.exit 0
end

taxes = [20, 10, 0]
cost_center_names = ['guest','restaurant','broken']
tax_colors = ['#e3bde1', '#f3d5ab','#ffafcf','#d2f694','#c2e8f3','#c6c6c6']
category_colors = ['#80477d','#ed8b00','#cd0052','#75b10d','#136880','#27343b']
user_colors = ['#80477d','#ed8b00','#cd0052','#75b10d','#136880','#27343b','#BBBBBB','#000000','#d9d43d','#801212']
vendor_printer_labels = ['Bar','Kitchen','Guestroom']
payment_method_names = ['Cash', 'Card', 'Other']
role_names = {
  'superuser' =>
    {:weight => 0, :permissions => ['take_orders','decrement_items','delete_items','cancel_all_items_in_active_order','finish_orders','split_items','move_tables','refund','assign_cost_center','assign_order_to_booking','move_order','manage_articles','manage_categories','manage_options','finish_all_settlements','finish_own_settlement','view_all_settlements','manage_business_invoice','manage_statistics','manage_users','manage_taxes','manage_cost_centers','manage_payment_methods','manage_tables','manage_vendors','counter_mode','see_item_notifications_delivery','see_item_notifications_preparation','see_item_notifications_vendor','manage_pages','manage_customers','see_debug','manage_hotel','manage_roles','item_scribe','assign_tables','download_database','remote_support']},
  'owner' =>
    {:weight => 1, :permissions => ['take_orders','decrement_items','delete_items','cancel_all_items_in_active_order','finish_orders','split_items','move_tables','refund','move_order','manage_articles','manage_categories','manage_users','manage_taxes','manage_tables','manage_vendors'] },
  'host' =>
    {:weight => 2, :permissions => ['take_orders','decrement_items','delete_items','cancel_all_items_in_active_order','finish_orders','split_items','move_tables','refund','move_order','manage_articles','manage_categories','manage_users','manage_taxes','manage_tables'] },
  'chief_waiter' =>
    {:weight => 3, :permissions => ['take_orders','decrement_items','delete_items','cancel_all_items_in_active_order','finish_orders','split_items','move_tables','refund','move_order','manage_articles','manage_tables'] },
  'waiter' =>
    {:weight => 4, :permissions => ['take_orders','decrement_items','finish_orders','split_items','move_order']},
  'auxiliary_waiter' =>
    {:weight => 5, :permissions => ['take_orders','finish_orders']},
  'terminal' =>
    {:weight => 6, :permissions => ['take_orders'] },
  'customer' =>
    {:weight => 10, :permissions => [] }
}

user_array = {
  'Superuser' => {:role => 'superuser'},
  'Owner' => {:role => 'owner'},
  'Host' => {:role => 'host'},
  'Chief Waiter' => {:role => 'chief_waiter'},
  'Waiter' => {:role => 'waiter'},
  'Auxiliary Waiter' => {:role => 'auxiliary_waiter'},
  'Terminal' => {:role => 'terminal' },
  'Customer' => {:role => 'customer' }
  }

radio_surcharge_names = ['Breakfast','Dinner','Full']
checkbox_surcharge_names = ['Additional Bed']
common_surcharge_names = ['1 Night', '2 Night']
surcharge_amounts = [6, 12, 22, -5]

Article.delete_all
BookingItem.delete_all
Booking.delete_all
CashDrawer.delete_all
CashRegister.delete_all
Category.delete_all
Company.delete_all
CostCenter.delete_all
Coupon.delete_all
Customer.delete_all
Discount.delete_all
Group.delete_all
GuestType.delete_all
History.delete_all
Image.delete_all
Ingredient.delete_all
Item.delete_all
Option.delete_all
Order.delete_all
Page.delete_all
Partial.delete_all
PaymentMethod.delete_all
Presentation.delete_all
Quantity.delete_all
Receipt.delete_all
Reservation.delete_all
Role.delete_all
RoomPrice.delete_all
RoomType.delete_all
Room.delete_all
Season.delete_all
Settlement.delete_all
Stock.delete_all
SurchargeItem.delete_all
Surcharge.delete_all
Table.delete_all
TaxAmount.delete_all
TaxItem.delete_all
Tax.delete_all
User.delete_all
VendorPrinter.delete_all
Vendor.delete_all



company_count.times do |c|
  company = Company.new :name => "Company #{ c }"
  r = company.save
  puts "Company #{ c } created" if r == true

  countries.size.times do |v|
    vendor = Vendor.new :name => "Vendor #{ c } #{ v }", :country => countries[v]
    vendor.company = company
    r = vendor.save
    puts "Vendor #{ c } #{ v } created" if r == true

    presentation_objects = Array.new
    presentation1 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "Article name #{ c } #{ v }", :description => "just displays name of an article", :markup => "%{{NAME}}", :model => "Article"
    presentation2 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "Variant list #{ c } #{ v }", :description => "displays names of variants belonging to an article", :markup => "%{{LOOP BEGIN QUANTITIES}}\r\n%{{LOOP.PREFIX}}<br/>\r\n%{{END}}", :model => "Article"
    presentation3 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "Generic Text Blurb #{ c } #{ v }", :description => "displays any entered text", :markup => "%{{BLURB}}", :model => "Presentation"
    presentation4 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "Category Title #{ c } #{ v }", :description => "just displays name of category", :markup => "%{{NAME}}", :model => "Category"
    presentation5 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "Article full #{ c } #{ v }", :description => "displays Article with name, price and image", :markup => "<b>%{{NAME}}</b><br />\r\nEUR %{{PRICE}}\r\n%{{IF IMAGE}}\r\n<br/>\r\n<img src=\"%{{IMAGE}}\">\r\n%{{END}}", :model => "Article"
    presentation6 = Presentation.create :vendor_id => vendor.id, :company_id => company.id, :name => "All articles of a category #{ c } #{ v }", :description => "displays All articles of a category, including descriptions", :markup => "<table width=300 border=0>\r\n%{{LOOP BEGIN ARTICLES}}\r\n<tr>\r\n<td><br/><b><big>%{{LOOP.NAME}}</big></b></td>\r\n<td width=100 rowspan=2 align=right><br/><i>EUR %{{LOOP.PRICE}}</i></td>\r\n</tr>\r\n<tr>\r\n<td><br/><small>%{{LOOP.DESCRIPTION}}</small></td>\r\n</tr>\r\n%{{END}}\r\n</table>", :model => "Category"
    presentation_objects << presentation1
    presentation_objects << presentation2
    presentation_objects << presentation3
    presentation_objects << presentation4
    presentation_objects << presentation5
    presentation_objects << presentation6
    puts "Presentations #{ c } #{ v } created"

    partial_objects = Array.new
    partial1 = Partial.create "left"=>230, "top"=>-19, "presentation_id"=>presentation3.id, "blurb"=>"Holiday Offer", "active"=>true, "hidden"=>nil, "model_id"=>4, "font"=>"Rolina", "size"=>400, "image_size"=>nil, "color"=>"#e67373", "width"=>nil, "align"=>nil, "company_id"=>company.id, "vendor_id"=>vendor.id
    partial2 = Partial.create "left"=>117, "top"=>187, "presentation_id"=>presentation6.id, "blurb"=>nil, "active"=>true, "hidden"=>nil, "model_id"=>1, "font"=>nil, "size"=>150, "image_size"=>250, "color"=>"#c3d3e6", "width"=>800, "align"=>nil, "company_id"=>company.id, "vendor_id"=>vendor.id
    partial3 = Partial.create "left"=>583, "top"=>242, "presentation_id"=>presentation2.id, "blurb"=>nil, "active"=>true, "hidden"=>nil, "model_id"=>2, "font"=>nil, "size"=>200, "image_size"=>nil, "color"=>"#95e6e1", "width"=>nil, "align"=>nil, "company_id"=>company.id, "vendor_id"=>vendor.id
    partial4 = Partial.create "left"=>246, "top"=>165, "presentation_id"=>presentation5.id, "blurb"=>nil, "active"=>true, "hidden"=>nil, "model_id"=>1, "font"=>"Rolina", "size"=>400, "image_size"=>nil, "color"=>"#7579c9", "width"=>nil, "align"=>"center", "company_id"=>company.id, "vendor_id"=>vendor.id
    partial5 = Partial.create "left"=>203, "top"=>21, "presentation_id"=>presentation4.id, "blurb"=>nil, "active"=>true, "hidden"=>nil, "model_id"=>1, "font"=>"aaaiight", "size"=>500, "image_size"=>nil, "color"=>"#5b3bdb", "width"=>nil, "align"=>nil, "company_id"=>company.id, "vendor_id"=>vendor.id
    partial_objects << partial1
    partial_objects << partial2
    partial_objects << partial3
    partial_objects << partial4
    partial_objects << partial5
    puts "Partials #{ c } #{ v } created"

    pages_objects = Array.new
    page1 = Page.create "color"=>"#1f1818", "company_id"=>company.id, "vendor_id"=>vendor.id
    page2 = Page.create "color"=>"#8adea7", "company_id"=>company.id, "vendor_id"=>vendor.id

    page1.partials << partial1
    page1.partials << partial2
    page1.partials << partial3
    page1.save

    page2.partials << partial4
    page2.partials << partial5
    page2.save
    puts "Pages #{ c } #{ v } created"


    tax_objects = Array.new
    letters = ['A','B','C']
    taxes.size.times do |i|
      tax = Tax.new :name => "#{ taxes[i] } #{ c } #{ v }", :letter => letters[i], :percent => taxes[i], :color => tax_colors[i]
      tax.company = company
      tax.vendor = vendor
      r = tax.save
      tax_objects << tax
      puts "Tax #{ taxes[i] } #{ c } #{ v } created" if r == true
    end

    cash_register_objects = Array.new
    2.times do |i|
      cash_register = CashRegister.new :name => "CashRegister #{ c } #{ v } #{ i }"
      cash_register.company = company
      cash_register.vendor = vendor
      r = cash_register.save
      cash_register_objects << cash_register
      puts "CashRegister #{ c } #{ v } #{ i } created" if r == true
    end

    vendor_printer_objects = Array.new
    vendor_printer_labels.size.times do |i|
      vendor_printer = VendorPrinter.new :name => vendor_printer_labels[i], :path => "/tmp/vendor_printer_#{ c }_#{ v }_#{ i }.txt"
      vendor_printer.company = company
      vendor_printer.vendor = vendor
      r = vendor_printer.save
      vendor_printer_objects << vendor_printer
      puts "VendorPrinter #{ c } #{ v } #{ i } created" if r == true
    end

    role_objects = Array.new
    role_names.to_a.size.times do |i|
      role = Role.new :name => role_names.to_a[i][0], :permissions => role_names.to_a[i][1][:permissions], :weight => role_names.to_a[i][1][:weight]
      role.company = company
      role.vendor = vendor
      r = role.save
      role_objects << role
      puts "Role #{ user_array.to_a[i][0] } #{ c } #{ v } #{ i } created" if r == true
    end

    table_objects = Array.new
    3.times do |i|
      table = Table.new :name => "T#{ c }#{ v }#{ i }", :left => 50 * i + 50, :top => 50 * i + 50, :left_mobile => 50 * i + 50, :top_mobile => 100 * i + 50, :width => 70, :height => 70
      table.company = company
      table.vendor = vendor
      table.booking_table = true if i == 3
      r = table.save
      table_objects << table
      puts "Table #{ c } #{ v } #{ i } created" if r == true
    end

    cost_center_objects = Array.new
    cost_center_names.size.times do |i|
      cost_center = CostCenter.new :name => "CC #{ c }#{ v }#{ i }", :active => true
      cost_center.company = company
      cost_center.vendor = vendor
      r = cost_center.save
      cost_center_objects << cost_center
      puts "Cost Center #{ c } #{ v } #{ i } created" if r == true
    end

    customer_objects = Array.new
    10.times do |i|
      customer = Customer.new :first_name => "Bob#{c}#{v}#{i}", :last_name => "Doe#{c}#{v}#{i}", :company_name => "Company#{c}#{v}#{i}", :address => "Address#{c}#{v}#{i}", :m_points => 100-i
      customer.company = company
      customer.vendor = vendor
      r = customer.save
      customer_objects << customer
      puts "Customer #{ c } #{ v } #{ i } created" if r == true
    end

    payment_method_objects = Array.new
    payment_method_names.size.times do |i|
      pm = PaymentMethod.new :name => "#{ payment_method_names[i] }#{c}#{v}#{i}"
      pm.company = company
      pm.vendor = vendor
      r = pm.save
      payment_method_objects << pm
      puts "PaymentMethod #{ c } #{ v } #{ i } created" if r == true
    end

    user_objects = Array.new
    user_array.to_a.size.times do |i|
      user = User.new :login => "#{ user_array.to_a[i][0] } #{ c } #{ v } #{ i }", :title => "#{ user_array.to_a[i][0] }", :password => "#{ c }#{ v }#{ i }", :language => languages[v], :color => user_colors[i]
      user.company = company
      user.vendors << vendor
      user.tables = table_objects
      user.role = role_objects[i]
      r = user.save
      user_objects << user
      puts "User #{ user_array.to_a[i][0] } #{ c } #{ v } #{ i } created" if r == true

      cash_drawer = CashDrawer.new :name => "CashDrawer #{ c } #{ v } #{ i }"
      cash_drawer.company = company
      cash_drawer.vendor = vendor
      cash_drawer.user = user
      r = cash_drawer.save
      puts "CashDrawer #{ c } #{ v } #{ i } created" if r == true
    end

    category_labels.size.times do |i|
      category = Category.new :name => category_labels[i], :icon => category_icons[i], :color => category_colors[rand(category_colors.size)]
      category.company = company
      category.vendor = vendor
      category.preparation_user_id = user_objects.first.id
      category.vendor_printer = vendor_printer_objects.first
      r = category.save
      puts "Category #{ c } #{ v } #{ i } created" if r == true

      2.times do |o|
        option = Option.new :name => "Option#{ c }#{ v }#{ i }#{ o }", :price => (rand(3) + 1).to_f
        category.options << option
        option.company = company
        option.vendor = vendor
        r = option.save
        puts "Option #{ c } #{ v } #{ i } #{ o } created" if r == true
      end
        
      2.times do |a|
        article = Article.new :name => "Article#{ c }#{ v }#{ i }#{ a }", :price => rand(30) + 1, :active => true
        article.taxes = [tax_objects[rand(3)]]
        article.category = category
        article.company = company
        article.vendor = vendor
        r = article.save
        puts "Article #{ c } #{ v } #{ i } #{ a } created" if r == true
        
        if a > 0
          2.times do |q|
            quantity = Quantity.new :prefix => "#{c}#{v}#{i}#{a}#{q}", :postfix => 'post', :price => rand(10) + 1
            quantity.article = article
            quantity.category = category
            quantity.company = company
            quantity.vendor = vendor
            r = quantity.save
            puts "Quantity #{ c } #{ v } #{ i } #{ a } #{ q } created" if r == true
          end
        end
      end
    end

    puts " Creating Hotel Tax #{c} #{v}"
    #local_tax = Tax.create :name => "Local Tax #{c} #{v}", :percent => 1, :vendor_id => vendor.id, :company_id => company.id, :letter => "D"
    room_type_objects = Array.new
    guest_type_objects = Array.new
    4.times do |i|
      puts " Creating RoomType #{c} #{v} #{i}"
      rt = RoomType.create :name => "RoomType #{c} #{v} #{i}", :vendor_id => vendor.id, :company_id => company.id
      room_type_objects << rt
      puts " Creating Room #{c} #{v} #{i}"
      Room.create :name => "Room#{i}", :room_type_id => rt.id, :vendor_id => vendor.id, :company_id => company.id
      puts " Creating GuestType #{c} #{v} #{i}"
      gt = GuestType.new :name => "GuestType #{c} #{v} #{i}", :vendor_id => vendor.id, :company_id => company.id
      gt.taxes << tax_objects[0]
      gt.save
      guest_type_objects << gt
    end
    puts " Creating Seasons for #{c} #{v}"
    s1 = Season.new :name => 'Summer', :from_date => Date.parse('2012-06-21'), :to_date => Date.parse('2012-09-20'), :vendor_id => vendor.id, :company_id => company.id
    s1.save
    s1.calculate_duration
    s2 = Season.new :name => 'Autumn', :from_date => Date.parse('2012-09-21'), :to_date => Date.parse('2012-12-20'), :vendor_id => vendor.id, :company_id => company.id
    s2.save
    s1.calculate_duration
    s3 = Season.new :name => 'Winter', :from_date => Date.parse('2012-12-21'), :to_date => Date.parse('2012-03-20'), :vendor_id => vendor.id, :company_id => company.id
    s3.save
    s1.calculate_duration
    s4 = Season.new :name => 'Spring', :from_date => Date.parse('2012-03-21'), :to_date => Date.parse('2012-06-20'), :vendor_id => vendor.id, :company_id => company.id
    s4.save
    s1.calculate_duration
    season_objects = [s1,s2,s3,s4]

    4.times do |i|
      4.times do |j|
        4.times do |k|
          puts " Creating RoomPrice #{c} #{v} #{i} #{j} #{k}"
          RoomPrice.create :season_id => season_objects[k].id, :room_type_id => room_type_objects[i].id, :guest_type_id => guest_type_objects[j].id, :base_price => (2 * i + 3 * j + 5 * k) * 10 + 3, :vendor_id => vendor.id, :company_id => company.id
        end
      end
    end

    season_objects.size.times do |x|
      guest_type_objects.size.times do |y|
        radio_surcharge_names.size.times do |z|
          puts " Creating Radio Surcharge #{radio_surcharge_names[z]}"
          surcharge = Surcharge.new :name => radio_surcharge_names[z], :vendor_id => vendor.id, :company_id => company.id, :season_id => season_objects[x].id, :guest_type_id => guest_type_objects[y].id, :radio_select => true
          tax_amount = TaxAmount.create :vendor_id => vendor.id, :company_id => company.id, :amount => 1 + x + y + z, :tax_id => tax_objects[rand(3)].id
          surcharge.tax_amounts = [tax_amount]
          surcharge.save
          surcharge.calculate_totals
        end
        checkbox_surcharge_names.size.times do |z|
          puts " Creating Checkbox Surcharge #{checkbox_surcharge_names[z]}"
          surcharge = Surcharge.create :name => checkbox_surcharge_names[z], :vendor_id => vendor.id, :company_id => company.id, :season_id => season_objects[x].id, :guest_type_id => guest_type_objects[y].id
          tax_amount = TaxAmount.create :vendor_id => vendor.id, :company_id => company.id, :amount => 1 + x + y + z, :tax_id => tax_objects[rand(3)].id
          surcharge.tax_amounts = [tax_amount]
          surcharge.save
          surcharge.calculate_totals
        end
      end
      common_surcharge_names.size.times do |z|
        puts " Creating Common Surcharge #{common_surcharge_names[z]}"
        surcharge = Surcharge.create :name => common_surcharge_names[z], :vendor_id => vendor.id, :company_id => company.id, :season_id => season_objects[x].id, :guest_type_id => nil
        tax_amount = TaxAmount.create :vendor_id => vendor.id, :company_id => company.id, :amount => 3 + x + z, :tax_id => tax_objects[rand(3)].id
        surcharge.tax_amounts = [tax_amount]
        surcharge.save
        surcharge.calculate_totals
      end
    end
    vendor.update_cache
  end
end
